semhash/semhash.py

```diff
@@ -309,133 +310,237 @@ def _validate_if_strings(self, records: Sequence[dict[str, str] | str]) -> Seque
-            budget=budget,
-            descending=descending,
+        results = self.index.query_top_k(self.index.vectors, k=100)
+
+        # Compute the average similarity for each record.
+        scores = [(record, np.mean(sims)) for record, (_, sims) in zip(dict_records, results)]
+        scores.sort(key=lambda x: x[1], reverse=descending)
+        selected = [record for record, _ in scores]
+        scores_selected = [sim for _, sim in scores]
+
+        ranking = FilterResult(
+            selected=selected,
+            filtered=[],
+            scores_selected=scores_selected,
+            scores_filtered=[],
+        )
+        self._ranking_cache = ranking
+        return ranking
+
+    def _mmr(
+        self,
+        ranked_results: FilterResult,
+        candidate_limit: int,
+        selection_size: int,
+        lambda_param: float,
+    ) -> FilterResult:
+        """
+        Perform Maximal Marginal Relevance (MMR) re-ranking on the top candidates from a FilterResult.
+
+        This function first slices the ranking, then computes embeddings for the candidates,
+        normalizes them, and finally performs MMR re-ranking to obtain a diverse subset of representatives.
+
+        :param ranked_results: A FilterResult containing the ranking of records.
+        :param candidate_limit: Number of top candidates to consider from the ranking.
+        :param selection_size: Number of final candidates to select using MMR.
+        :param lambda_param: Weight balancing relevance and diversity (between 0 and 1).
+        :return: A FilterResult containing the selected (diversified) representatives along with
+                their MMR scores in `scores_selected`. The remaining candidates are placed in filtered.
+        """
+        # Slice the top candidates from the ranking.
+        candidate_records = ranked_results.selected[:candidate_limit]
+        candidate_relevance = ranked_results.scores_selected[:candidate_limit]
+
+        # Compute embeddings for candidate records.
+        embeddings = self._featurize(records=candidate_records, columns=self.columns, model=self.model)
+
+        # Normalize embeddings for cosine similarity.
+        norm_embeddings = []
+        for vec in embeddings:
+            norm = np.linalg.norm(vec)
+            norm_embeddings.append(vec / norm if norm != 0 else vec)
+
+        # Package candidates as tuples: (record, baseline_relevance, normalized_embedding)
+        candidates = list(zip(candidate_records, candidate_relevance, np.array(norm_embeddings)))
+
+        # MMR re-ranking: Initialize selected set with the candidate having highest baseline relevance.
+        selected_records = []
+        selected_scores = []  # MMR scores for selected candidates.
+        selected_embeddings = []  # For computing similarity with new candidates.
+        remaining_candidates = candidates.copy()
+
+        if remaining_candidates:
+            first = max(remaining_candidates, key=lambda x: x[1])
```

Comment:
isn't the first item always the one with the largest relevance? Since they are ranked by relevance.